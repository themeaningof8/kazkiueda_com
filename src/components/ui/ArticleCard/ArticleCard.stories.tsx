import type { Meta, StoryObj } from '@storybook/react'
import { ArticleCard } from './ArticleCard'

const meta: Meta<typeof ArticleCard> = {
  title: 'UI/ArticleCard',
  component: ArticleCard,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    categoryVariant: {
      control: 'select',
      options: ['default', 'secondary', 'destructive', 'outline'],
    },
    isExternal: {
      control: 'boolean',
    },
    layout: {
      control: 'select',
      options: ['vertical', 'horizontal'],
    },
    aspectRatio: {
      control: { type: 'number', min: 0.1, max: 5, step: 0.1 },
    },
  },
}

export default meta
type Story = StoryObj<typeof meta>

export const Default: Story = {
  args: {
    title: 'React 18の新機能を詳しく解説',
    category: 'フロントエンド',
    imageUrl: 'https://images.unsplash.com/photo-1633356122544-f134324a6cee?ixlib=rb-4.0.3&auto=format&fit=crop&w=400&h=200&q=80',
    href: '/articles/react-18-features',
    description: 'React 18で追加された新しい機能について、実際のコード例と共に詳しく解説します。',
  },
}

export const WithPublicImage: Story = {
  args: {
    title: 'TypeScriptベストプラクティス',
    category: '開発',
    imageUrl: '/placeholder-image.jpg', // publicディレクトリの画像例
    href: '/articles/typescript-best-practices',
    description: 'TypeScriptを使った開発でのベストプラクティスをまとめました。',
    categoryVariant: 'outline',
  },
}

export const ExternalLink: Story = {
  args: {
    title: 'Qiitaで書いた記事',
    category: 'エンジニアリング',
    imageUrl: 'https://images.unsplash.com/photo-1551650975-87deedd944c3?ixlib=rb-4.0.3&auto=format&fit=crop&w=400&h=200&q=80',
    href: 'https://qiita.com/example',
    isExternal: true,
    description: 'Qiitaに投稿した技術記事です。外部リンクとして開きます。',
    categoryVariant: 'default',
  },
}

export const LongTitle: Story = {
  args: {
    title: 'とても長いタイトルの記事：モダンなWebアプリケーション開発における設計パターンとアーキテクチャの選択について',
    category: 'アーキテクチャ',
    imageUrl: 'https://images.unsplash.com/photo-1558494949-ef010cbdcc31?ixlib=rb-4.0.3&auto=format&fit=crop&w=400&h=200&q=80',
    href: '/articles/web-app-architecture',
    description: 'モダンなWebアプリケーション開発で重要な設計パターンとアーキテクチャの選択について、実際のプロジェクト経験を基に解説します。',
    categoryVariant: 'secondary',
  },
}

export const NoDescription: Story = {
  args: {
    title: '説明なしの記事カード',
    category: 'シンプル',
    imageUrl: 'https://images.unsplash.com/photo-1498050108023-c5249f4df085?ixlib=rb-4.0.3&auto=format&fit=crop&w=400&h=200&q=80',
    href: '/articles/simple-card',
    categoryVariant: 'destructive',
  },
}

export const CustomClassName: Story = {
  args: {
    title: 'カスタムスタイル適用例',
    category: 'デザイン',
    imageUrl: 'https://images.unsplash.com/photo-1581291518857-4e27b48ff24e?ixlib=rb-4.0.3&auto=format&fit=crop&w=400&h=200&q=80',
    href: '/articles/custom-style',
    description: 'カスタムクラス名を適用した記事カードの例です。',
  },
}

export const HorizontalLayout: Story = {
  args: {
    title: '横並びレイアウトの記事カード',
    category: 'レイアウト',
    imageUrl: 'https://images.unsplash.com/photo-1498050108023-c5249f4df085?ixlib=rb-4.0.3&auto=format&fit=crop&w=400&h=200&q=80',
    href: '/articles/horizontal-layout',
    description: '画像とテキストが横並びで表示される記事カードのレイアウトです。コンパクトな表示に適しています。',
    layout: 'horizontal',
    aspectRatio: 1,
    categoryVariant: 'outline',
  },
}

// アスペクト比のバリエーション
export const AspectRatioVariations: Story = {
  render: () => (
    <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6 p-6 max-w-6xl">
      <ArticleCard
        title="正方形カード (1:1)"
        category="デザイン"
        imageUrl="https://images.unsplash.com/photo-1618005182384-a83a8bd57fbe?ixlib=rb-4.0.3&auto=format&fit=crop&w=400&h=400&q=80"
        href="/articles/square-design"
        description="正方形のアスペクト比で表示される記事カードです。"
        aspectRatio={1}
      />
      <ArticleCard
        title="ポートレート (3:4)"
        category="写真"
        imageUrl="https://images.unsplash.com/photo-1506905925346-21bda4d32df4?ixlib=rb-4.0.3&auto=format&fit=crop&w=300&h=400&q=80"
        href="/articles/portrait-photography"
        description="ポートレート比率で縦長に表示される記事カードです。"
        aspectRatio={3 / 4}
        categoryVariant="outline"
      />
      <ArticleCard
        title="ウルトラワイド (21:9)"
        category="映像"
        imageUrl="https://images.unsplash.com/photo-1440404653325-ab127d49abc1?ixlib=rb-4.0.3&auto=format&fit=crop&w=700&h=300&q=80"
        href="/articles/ultra-wide-video"
        description="ウルトラワイド比率で横長に表示される記事カードです。"
        aspectRatio={21 / 9}
        categoryVariant="default"
      />
    </div>
  ),
  parameters: {
    layout: 'fullscreen',
  },
}

// レイアウト比較
export const LayoutComparison: Story = {
  render: () => (
    <div className="space-y-8 p-6 max-w-4xl">
      <div>
        <h3 className="text-lg font-semibold mb-4">縦レイアウト（Vertical）</h3>
        <div className="grid grid-cols-1 md:grid-cols-2 gap-4">
          <ArticleCard
            title="縦レイアウトの記事カード例"
            category="デザイン"
            imageUrl="https://images.unsplash.com/photo-1558494949-ef010cbdcc31?ixlib=rb-4.0.3&auto=format&fit=crop&w=400&h=200&q=80"
            href="/articles/vertical-example"
            description="縦方向のレイアウトで、画像が上部に配置される従来のカードスタイルです。"
            layout="vertical"
          />
          <ArticleCard
            title="もう一つの縦レイアウト例"
            category="開発"
            imageUrl="https://images.unsplash.com/photo-1516321318423-f06f85e504b3?ixlib=rb-4.0.3&auto=format&fit=crop&w=400&h=200&q=80"
            href="/articles/vertical-example-2"
            description="画像とコンテンツが縦に並ぶ、読みやすいレイアウトです。"
            layout="vertical"
            categoryVariant="outline"
          />
        </div>
      </div>
      
      <div>
        <h3 className="text-lg font-semibold mb-4">横レイアウト（Horizontal）</h3>
        <div className="space-y-4">
          <ArticleCard
            title="横並びレイアウトの記事カード"
            category="UI/UX"
            imageUrl="https://images.unsplash.com/photo-1498050108023-c5249f4df085?ixlib=rb-4.0.3&auto=format&fit=crop&w=400&h=200&q=80"
            href="/articles/horizontal-example"
            description="画像とテキストが横並びになる、コンパクトで効率的なレイアウトです。リスト表示に適しています。"
            layout="horizontal"
            aspectRatio={1}
          />
          <ArticleCard
            title="効率的な情報表示を実現する横レイアウト"
            category="エンジニアリング"
            imageUrl="https://images.unsplash.com/photo-1551650975-87deedd944c3?ixlib=rb-4.0.3&auto=format&fit=crop&w=400&h=200&q=80"
            href="/articles/horizontal-example-2"
            description="限られたスペースで多くの情報を表示できる横型レイアウトの活用方法について解説します。"
            layout="horizontal"
            aspectRatio={4/3}
            categoryVariant="default"
          />
        </div>
      </div>
    </div>
  ),
  parameters: {
    layout: 'fullscreen',
  },
}

// グリッドレイアウトでの表示例
export const GridLayout: Story = {
  render: () => (
    <div className="grid grid-cols-1 md:grid-cols-2 lg:grid-cols-3 gap-6 p-6 max-w-6xl">
      <ArticleCard
        title="React Hooksの使い方"
        category="React"
        imageUrl="https://images.unsplash.com/photo-1516321318423-f06f85e504b3?ixlib=rb-4.0.3&auto=format&fit=crop&w=400&h=200&q=80"
        href="/articles/react-hooks"
        description="React Hooksの基本的な使い方を解説します。"
      />
      <ArticleCard
        title="Next.js 13 App Routerの新機能"
        category="Next.js"
        imageUrl="https://images.unsplash.com/photo-1555066931-4365d14bab8c?ixlib=rb-4.0.3&auto=format&fit=crop&w=400&h=200&q=80"
        href="/articles/nextjs-app-router"
        description="Next.js 13のApp Routerで追加された新機能について。"
        categoryVariant="outline"
      />
      <ArticleCard
        title="Tailwind CSSのベストプラクティス"
        category="CSS"
        imageUrl="https://images.unsplash.com/photo-1507003211169-0a1dd7228f2d?ixlib=rb-4.0.3&auto=format&fit=crop&w=400&h=200&q=80"
        href="/articles/tailwind-best-practices"
        description="Tailwind CSSを効率的に使うためのベストプラクティス。"
        categoryVariant="default"
      />
    </div>
  ),
  parameters: {
    layout: 'fullscreen',
  },
} 