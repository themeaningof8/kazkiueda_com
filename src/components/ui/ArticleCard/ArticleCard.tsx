import React from 'react'

import { AspectRatio } from '@/components/ui/AspectRatio'
import { Badge } from '@/components/ui/Badge'
import { Card, CardContent, CardHeader } from '@/components/ui/Card'
import { cn } from '@/utils'

export interface ArticleCardProps {
  /** 記事のタイトル */
  title: string
  /** カテゴリ名 */
  category: string
  /** 画像のURL（外部URLまたはpublic内のパス） */
  imageUrl: string
  /** 記事のリンク先URL */
  href: string
  /** 記事の説明（オプション） */
  description?: string
  /** 画像のalt属性 */
  imageAlt?: string
  /** 外部リンクかどうか */
  isExternal?: boolean
  /** 追加のクラス名 */
  className?: string
  /** カテゴリのバリアント */
  categoryVariant?: 'default' | 'secondary' | 'destructive' | 'outline'
  /** 画像のアスペクト比 */
  aspectRatio?: number
  /** レイアウト */
  layout?: 'vertical' | 'horizontal'
}

export const ArticleCard: React.FC<ArticleCardProps> = ({
  title,
  category,
  imageUrl,
  href,
  description,
  imageAlt,
  isExternal = false,
  className,
  categoryVariant = 'secondary',
  aspectRatio = 21 / 9,
  layout = 'vertical',
}) => {
  const linkProps = isExternal ? { target: '_blank', rel: 'noopener noreferrer' } : {}
  const isHorizontal = layout === 'horizontal'

  return (
    <Card className={cn('group overflow-hidden transition-all pt-0 hover:shadow-lg', className)}>
      <a
        href={href}
        {...linkProps}
        className={cn('block space-y-4', isHorizontal && 'flex items-start gap-4')}
      >
        <CardHeader className='p-0'>
          <AspectRatio
            ratio={aspectRatio}
            className={cn('relative w-full overflow-hidden', isHorizontal && 'w-32 flex-shrink-0')}
          >
            <img
              src={imageUrl}
              alt={imageAlt || title}
              className='h-full w-full object-cover transition-transform group-hover:scale-105'
              loading='lazy'
            />
            <div className='absolute inset-0 bg-black/0 transition-colors group-hover:bg-black/10' />
          </AspectRatio>
        </CardHeader>
        <CardContent className={cn('space-y-2', isHorizontal && 'flex-grow')}>
          <h3 className='mb-2 font-semibold leading-tight group-hover:text-primary transition-colors'>
            {title}
          </h3>
          {description && (
            <p className='text-sm text-muted-foreground line-clamp-2'>{description}</p>
          )}
          <div className='flex items-center justify-end'>
            <Badge variant={categoryVariant} className='text-xs'>
              {category}
            </Badge>
          </div>
        </CardContent>
      </a>
    </Card>
  )
}
