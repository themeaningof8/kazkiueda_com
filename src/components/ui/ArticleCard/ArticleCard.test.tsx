import { render, screen } from '@testing-library/react'
import { describe, it, expect } from 'vitest'
import { ArticleCard } from './ArticleCard'

const defaultProps = {
  title: 'テスト記事タイトル',
  category: 'テクノロジー',
  imageUrl: 'https://images.unsplash.com/photo-1516321318423-f06f85e504b3?ixlib=rb-4.0.3&auto=format&fit=crop&w=400&h=200&q=80',
  href: '/test-article',
}

describe('ArticleCard', () => {
  it('必須プロパティが正しく表示される', () => {
    render(<ArticleCard {...defaultProps} />)
    
    expect(screen.getByText('テスト記事タイトル')).toBeInTheDocument()
    expect(screen.getByText('テクノロジー')).toBeInTheDocument()
    expect(screen.getByRole('img')).toHaveAttribute('src', defaultProps.imageUrl)
    expect(screen.getByRole('link')).toHaveAttribute('href', defaultProps.href)
  })

  it('説明が提供された時に表示される', () => {
    const description = 'これはテスト記事の説明です。'
    render(<ArticleCard {...defaultProps} description={description} />)
    
    expect(screen.getByText(description)).toBeInTheDocument()
  })

  it('外部リンクの場合に適切な属性が設定される', () => {
    render(<ArticleCard {...defaultProps} isExternal={true} />)
    
    const link = screen.getByRole('link')
    expect(link).toHaveAttribute('target', '_blank')
    expect(link).toHaveAttribute('rel', 'noopener noreferrer')
  })
  it('画像のalt属性が設定される', () => {
    const altText = 'カスタムalt属性'
    render(<ArticleCard {...defaultProps} imageAlt={altText} />)
    
    expect(screen.getByRole('img')).toHaveAttribute('alt', altText)
  })

  it('alt属性が提供されない場合、タイトルがalt属性に使用される', () => {
    render(<ArticleCard {...defaultProps} />)
    
    expect(screen.getByRole('img')).toHaveAttribute('alt', defaultProps.title)
  })

  it('異なるカテゴリバリアントが適用される', () => {
    render(<ArticleCard {...defaultProps} categoryVariant="outline" />)
    
    const badge = screen.getByText('テクノロジー')
    expect(badge).toBeInTheDocument()
  })

  it('publicディレクトリの画像URLで動作する', () => {
    const publicImageUrl = '/images/test-image.jpg'
    render(<ArticleCard {...defaultProps} imageUrl={publicImageUrl} />)
    
    expect(screen.getByRole('img')).toHaveAttribute('src', publicImageUrl)
  })

  it('縦レイアウト（デフォルト）で正しく表示される', () => {
    render(<ArticleCard {...defaultProps} layout="vertical" />)
    
    const link = screen.getByRole('link')
    expect(link).toHaveClass('flex-col')
  })

  it('横レイアウトで正しく表示される', () => {
    render(<ArticleCard {...defaultProps} layout="horizontal" />)
    
    const link = screen.getByRole('link')
    expect(link).not.toHaveClass('flex-col')
    expect(link).toHaveClass('flex')
  })

  it('カスタムアスペクト比が適用される', () => {
    render(<ArticleCard {...defaultProps} aspectRatio={1} />)
    
    // AspectRatioコンポーネントが正しく表示されることを確認
    expect(screen.getByRole('img')).toBeInTheDocument()
  })

  it('異なるUnsplash画像で動作する', () => {
    const unsplashImageUrl = 'https://images.unsplash.com/photo-1551650975-87deedd944c3?ixlib=rb-4.0.3&auto=format&fit=crop&w=400&h=200&q=80'
    render(<ArticleCard {...defaultProps} imageUrl={unsplashImageUrl} />)
    
    expect(screen.getByRole('img')).toHaveAttribute('src', unsplashImageUrl)
  })

}) 