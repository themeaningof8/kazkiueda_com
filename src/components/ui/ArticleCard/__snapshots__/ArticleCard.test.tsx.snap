// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ArticleCard > 横レイアウトで正しく表示される 1`] = `
<DocumentFragment>
  <div
    class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm group overflow-hidden transition-all pt-0 hover:shadow-lg"
    data-slot="card"
  >
    <a
      class="space-y-4 flex items-start gap-4"
      href="/test-article"
    >
      <div
        class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 p-0"
        data-slot="card-header"
      >
        <div
          data-radix-aspect-ratio-wrapper=""
          style="position: relative; width: 100%; padding-bottom: 42.857142857142854%;"
        >
          <div
            class="relative overflow-hidden w-32 flex-shrink-0"
            data-slot="aspect-ratio"
            style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px;"
          >
            <img
              alt="テスト記事タイトル"
              class="h-full w-full object-cover transition-transform group-hover:scale-105"
              loading="lazy"
              src="https://images.unsplash.com/photo-1516321318423-f06f85e504b3?ixlib=rb-4.0.3&auto=format&fit=crop&w=400&h=200&q=80"
            />
            <div
              class="absolute inset-0 bg-black/0 transition-colors group-hover:bg-black/10"
            />
          </div>
        </div>
      </div>
      <div
        class="px-6 space-y-2 flex-grow"
        data-slot="card-content"
      >
        <h3
          class="mb-2 font-semibold leading-tight group-hover:text-primary transition-colors"
        >
          テスト記事タイトル
        </h3>
        <div
          class="flex items-center justify-end"
        >
          <span
            class="inline-flex items-center justify-center rounded-md border px-2 py-0.5 font-medium w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden border-transparent bg-secondary text-secondary-foreground [a&]:hover:bg-secondary/90 text-xs"
            data-slot="badge"
          >
            テクノロジー
          </span>
        </div>
      </div>
    </a>
  </div>
</DocumentFragment>
`;

exports[`ArticleCard > 縦レイアウト（デフォルト）で正しく表示される 1`] = `
<DocumentFragment>
  <div
    class="bg-card text-card-foreground flex flex-col gap-6 rounded-xl border py-6 shadow-sm group overflow-hidden transition-all pt-0 hover:shadow-lg"
    data-slot="card"
  >
    <a
      class="block space-y-4"
      href="/test-article"
    >
      <div
        class="@container/card-header grid auto-rows-min grid-rows-[auto_auto] items-start gap-1.5 has-data-[slot=card-action]:grid-cols-[1fr_auto] [.border-b]:pb-6 p-0"
        data-slot="card-header"
      >
        <div
          data-radix-aspect-ratio-wrapper=""
          style="position: relative; width: 100%; padding-bottom: 42.857142857142854%;"
        >
          <div
            class="relative w-full overflow-hidden"
            data-slot="aspect-ratio"
            style="position: absolute; top: 0px; right: 0px; bottom: 0px; left: 0px;"
          >
            <img
              alt="テスト記事タイトル"
              class="h-full w-full object-cover transition-transform group-hover:scale-105"
              loading="lazy"
              src="https://images.unsplash.com/photo-1516321318423-f06f85e504b3?ixlib=rb-4.0.3&auto=format&fit=crop&w=400&h=200&q=80"
            />
            <div
              class="absolute inset-0 bg-black/0 transition-colors group-hover:bg-black/10"
            />
          </div>
        </div>
      </div>
      <div
        class="px-6 space-y-2"
        data-slot="card-content"
      >
        <h3
          class="mb-2 font-semibold leading-tight group-hover:text-primary transition-colors"
        >
          テスト記事タイトル
        </h3>
        <div
          class="flex items-center justify-end"
        >
          <span
            class="inline-flex items-center justify-center rounded-md border px-2 py-0.5 font-medium w-fit whitespace-nowrap shrink-0 [&>svg]:size-3 gap-1 [&>svg]:pointer-events-none focus-visible:border-ring focus-visible:ring-ring/50 focus-visible:ring-[3px] aria-invalid:ring-destructive/20 dark:aria-invalid:ring-destructive/40 aria-invalid:border-destructive transition-[color,box-shadow] overflow-hidden border-transparent bg-secondary text-secondary-foreground [a&]:hover:bg-secondary/90 text-xs"
            data-slot="badge"
          >
            テクノロジー
          </span>
        </div>
      </div>
    </a>
  </div>
</DocumentFragment>
`;
